compute/computer.go

```diff
@@ -177,17 +185,222 @@ func (c *Computer) ClearCacheForProject(projectID uuid.UUID) {
 	c.cache.Delete(projectID)
 }
 
-func newValueHandler(states AccountStates) func(owner flow.Address, key string, value cadence.Value) error {
-	return func(owner flow.Address, key string, value cadence.Value) error {
-		// TODO: Remove address conversion
-		address := model.NewAddressFromBytes(owner.Bytes())
+func (c *Computer) extractStateChangesFromDelta(d delta.Delta, p *programs.Programs) AccountStates {
+
+	runtimeContext := runtime.Context{Interface: &apiEnv{
+		Delta:    &d,
+		Programs: p,
+	}}
+
+	getStored := func(address common.Address, key string) (value cadence.Value, err error) {
+		pathParts := strings.Split(key, "\x1F")
+
+		if len(pathParts) != 2 {
+			// Not a cadence path value. Safe to ignore.
+			return nil, nil
+		}
+
+		path := cadence.Path{
+			Domain:     pathParts[0],
+			Identifier: pathParts[1],
+		}
+
+		domain := common.PathDomainFromIdentifier(path.Domain)
+		if domain == common.PathDomainUnknown {
+			// Not a cadence path value. Safe to ignore.
+			return nil, nil
+		}
+
+		defer func() {
+			if r := recover(); r != nil {
+				// Something went wrong, could be that this isn't a cadence value.
+				c.logger.Debug().
+					Err(r.(error)).
+					Str("address", address.Hex()).
+					Str("key", key).
+					Msgf("error decoding value")
+
+				value = nil
+				err = nil
+			}
+		}()
+
+		return c.vm.Runtime.ReadStored(address, path, runtimeContext)
+	}
+
+	states := make(AccountStates)
+
+	ids, _ := d.RegisterUpdates()
+
+	for _, id := range ids {
+		addressBytes := []byte(id.Owner)
+		if len(addressBytes) != flow.AddressLength {
+			continue
+		}
+		commonAddress := common.BytesToAddress(addressBytes)
+		modelAddress := model.NewAddressFromBytes(addressBytes)
+
+		value, err := getStored(commonAddress, id.Key)
 
-		if _, ok := states[address]; !ok {
-			states[address] = make(map[string]cadence.Value)
+		if err != nil || value == nil {
+			// Not a cadence value or problem getting value.
+			continue
 		}
 
-		states[address][key] = value
+		if states[modelAddress] == nil {
+			states[modelAddress] = make(model.AccountState)
+		}
 
-		return nil
+		states[modelAddress][id.Key] = value
 	}
+
+	return states
+}
+
+var _ runtime.Interface = &apiEnv{}
+
+type apiEnv struct {
```

Comment:
Maybe add a comment here describing the purpose of this `runtime.Interface` implementation:
```suggestion
// apiEnv implements runtime.Interface solely for the purpose 
// of extracting state from a delta (see Computer.extractStateChangesFromDelta)
//
type apiEnv struct {
```